piglot/solver/crate/solver.py

```diff
@@ -0,0 +1,198 @@
+"""Module for CRATE solver."""
+from typing import Dict, Any, Type
+import os
+import time
+import shutil
+import subprocess
+from multiprocessing.pool import ThreadPool as Pool
+import numpy as np
+from piglot.parameter import ParameterSet
+from piglot.solver.solver import Solver, Case, CaseResult, OutputField, OutputResult, InputData
+from piglot.solver.crate.fields import crate_fields_reader, CRATEInputData
+from piglot.utils.solver_utils import has_keyword, load_module_from_file
+
+
+class CRATESolver(Solver):
+    """CRATE solver."""
+
+    def __init__(
+            self,
+            cases: Dict[Case, Dict[str, OutputField]],
+            parameters: ParameterSet,
+            output_dir: str,
+            crate_bin: str,
+            parallel: int,
+            tmp_dir: str,
+            ) -> None:
+        """Constructor for the CRATE solver class.
+
+        Parameters
+        ----------
+        cases : Dict[Case, Dict[str, OutputField]]
+            Cases to be run and respective output fields.
+        parameters : ParameterSet
+            Parameter set for this problem.
+        output_dir : str
+            Path to the output directory.
+        crate_bin : str
+            Path to the main.py CRATE file.
+        parallel : int
+            Number of parallel processes to use.
+        tmp_dir : str
+            Path to the temporary directory.
+        """
+        super().__init__(cases, parameters, output_dir)
+        self.crate_bin = crate_bin
+        self.parallel = parallel
+        self.tmp_dir = tmp_dir
```

Comment:
I believe we should also add an option for passing the discretisation directories.